packages/runtime/src/components/page/PageHead.tsx

```diff
@@ -0,0 +1,294 @@
+          <meta name="twitter:title" content={title} data-makeswift-meta="true" />
+          <meta itemProp="name" content={title} data-makeswift-meta="true" />
+        </>
+      )}
+
+      {description && (
+        <>
+          <meta name="description" content={description} data-makeswift-meta="true" />
+          <meta property="og:description" content={description} data-makeswift-meta="true" />
+          <meta name="twitter:description" content={description} data-makeswift-meta="true" />
+          <meta itemProp="description" content={description} data-makeswift-meta="true" />
+        </>
+      )}
+
+      {keywords && <meta name="keywords" content={keywords} data-makeswift-meta="true" />}
+
+      {socialImage && (
+        <>
+          <meta property="og:image" content={socialImage.publicUrl} data-makeswift-meta="true" />
+          <meta
+            property="og:image:type"
+            content={socialImage.mimetype}
+            data-makeswift-meta="true"
+          />
+          <meta name="twitter:image" content={socialImage.publicUrl} data-makeswift-meta="true" />
+          <meta name="twitter:card" content="summary_large_image" data-makeswift-meta="true" />
+          <meta itemProp="image" content={socialImage.publicUrl} data-makeswift-meta="true" />
+        </>
+      )}
+
+      {fontFamilyParamValue !== '' && (
+        <link
+          rel="stylesheet"
+          href={`https://fonts.googleapis.com/css?family=${fontFamilyParamValue}&display=swap`}
+          data-makeswift-meta="true"
+        />
+      )}
+
+      {headSnippets.map(snippetToElement).map(children =>
+        Children.map(children, child => {
+          if (typeof child === 'string') return child
+
+          if ((child.type as string) === 'script') {
+            return cloneElement(child, { ref: saveScriptRef })
+          }
+
+          if (VALID_HEAD_ELEMENT_TYPES.includes(child.type as string)) return child
+
+          return null
+        }),
+      )}
+    </>
+  )
+
+  // Won't inject head tag if in pages router (useServerInsertedHTML is a no-op)
+  // The ref is used to prevent a duplicate insertion:
+  // https://github.com/vercel/next.js/discussions/49354
+  const insertedServerHTML = useRef(false)
+  useServerInsertedHTML(() => {
+    if (insertedServerHTML.current) return
+    insertedServerHTML.current = true
+    return headElementsFragment
+  })
+
+  const isPagesRouter = useIsPagesRouter()
+  const [shouldRenderPortal, setShouldRenderPortal] = useState(false)
+
+  // Purge the initial head elements that were server inserted. Then,
+  // we can portal in the same head elements so they will now be tracked on
+  // client side navigation by React. Will not occur in pages router.
+  useEffect(() => {
+    if (isPagesRouter) return
+    document.head.querySelectorAll('[data-makeswift-meta]').forEach(el => el.remove())
+    setShouldRenderPortal(true)
+  }, [isPagesRouter])
```

Comment:
So my understanding of what's going on here is that we mark the server-inserted HTML elements with the `data-*` attribute and then we remove those _once_ and moving forward the portal takes care of client-side navigation. Is that accurate?

Reply:
Bingo